### PR TITLE
Add analyzer/fixer to suggest changing code like `ImmutableArray.Create(1, 2, 3)` to `[1, 2, 3]`

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Part of https://github.com/dotnet/roslyn/issues/69132

Followup to https://github.com/dotnet/roslyn/pull/69452

### CODE REVIEW

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs [hunk 1]
```diff
@@ -403,14 +404,19 @@ void AddCommaIfMissing()
                     return CreateCollectionElement(match, expression);
                 }
             }
+            else if (node is ExpressionSyntax expression)
+            {
+                // expression.GetAncestor<StatementSyntax>()
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs:409]:
```suggestion
```

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -91,7 +91,7 @@ private static void AnalyzeArrayCreationExpression(SyntaxNodeAnalysisContext con
         ReportArrayCreationDiagnostics(context, syntaxTree, option, arrayCreationExpression);
     }
 
-    public static ImmutableArray<CollectionExpressionMatch> TryGetMatches(
+    public static ImmutableArray<CollectionExpressionMatch<StatementSyntax>> TryGetMatches(
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:94]: this type became generic as we don't have statements when converting expressions (like `ImmutableArray.Create(1, 2, 3)`) over to collection literals.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs [hunk 3]
```diff
@@ -27,15 +27,16 @@ internal static class CSharpCollectionExpressionRewriter
     /// Creates the final collection-expression <c>[...]</c> that will replace the given <paramref
     /// name="expressionToReplace"/> expression.
     /// </summary>
-    public static async Task<CollectionExpressionSyntax> CreateCollectionExpressionAsync<TParentExpression>(
+    public static async Task<CollectionExpressionSyntax> CreateCollectionExpressionAsync<TParentExpression, TMatchNode>(
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs:30]: fallout of making the match type generic.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs [hunk 4]
```diff
@@ -569,12 +577,10 @@ string GetIndentationStringForToken(SyntaxToken token)
...
-            var tokenLine = document.Text.Lines.GetLineFromPosition(position);
-            var indentation = position - tokenLine.Start;
-            var indentationString = new IndentationResult(indentation, offset: 0).GetIndentationString(
-                document.Text, indentationOptions);
-
-            return indentationString;
+            var lineContainingPosition = document.Text.Lines.GetLineFromPosition(position);
+            var lineText = lineContainingPosition.ToString();
+            var indentation = lineText.ConvertTabToSpace(formattingOptions.TabSize, initialColumn: 0, endPosition: position - lineContainingPosition.Start);
+            return indentation.CreateIndentationString(formattingOptions.UseTabs, formattingOptions.TabSize);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs:583]: wasn't handling tabs properly before.  Will be adding tests on this as well.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForCreateCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        CancellationToken cancellationToken)
+    {
+        var unwrapArgument = properties.ContainsKey(CSharpUseCollectionExpressionForCreateDiagnosticAnalyzer.UnwrapArgument);
+
+        // We want to replace `XXX.Create(...)` with the new collection expression.  To do this, we go through the
+        // following steps.  First, we replace `XXX.Create(a, b, c)` with `new(a, b, c)` (an dummy object creation
+        // expression). We then call into our helper which replaces expressions with collection expressions.  The reason
+        // for the dummy object creation expression is that it serves as an actual node the rewriting code can attach an
+        // initializer to, by which it can figure out appropriate wrapping and indentation for the collection expression
+        // elements.
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpUseCollectionExpressionForCreateCodeFixProvider.cs:55]: a really funky way to do things.  but it allows us to leverage as much rewriting code as possible.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingExtensions.cs [hunk 6]
```diff
@@ -102,6 +102,7 @@ public static bool IsMaskOn(this SuppressOption option, SuppressOption mask)
         public static SuppressOption RemoveFlag(this SuppressOption option, SuppressOption flag)
             => option & ~flag;
 
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingExtensions.cs:105]:
```suggestion
```

**Comment by CyrusNajmabadi** [comment]
@akhera99 this is ready for review.  thanks!

**Review by akhera99 — APPROVED** [review]

**Pull request merged**